src/lib/staged_ledger/staged_ledger.ml

```diff
@@ -1777,6 +1777,16 @@ module T = struct
       epoch_ledger
     |> not
 
+  let with_ledger_mask base_ledger ~f =
```

Comment:
nice if this wasn't duplicated ...

Reply:
The control flow for this is different though. Here the mask doesn't get committed.

Reply:
We need 2 primitives here. One that's `ledger -> f:(ledger -> 'a) -> 'a` which never commits, and one that's `ledger -> f:(ledger -> ('a, 'b) result) -> ('a, 'b) result` which only commits if the return value of `f` is `Ok`.

Reply:
Could there be a function with an optional `~commit_result` flag?

Reply:
OK. It would be nice if these were generally available.

Reply:
Let's address this in a followup PR. It's not immediately obvious to me what the correct pattern is to abstract over these generally, and I think it's worth taking the time to factor these primitives into the rest of the code as well, so we can do that there.